osu.Game/Screens/Select/BeatmapDetails.cs

```diff
@@ -0,0 +1,430 @@
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using OpenTK;
+using OpenTK.Graphics;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Database;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using System.Globalization;
+using System.Linq;
+
+namespace osu.Game.Screens.Select
+{
+    public class BeatmapDetails : Container
+    {
+        private readonly MetadataSegment description;
+        private readonly MetadataSegment source;
+        private readonly MetadataSegment tags;
+
+        private readonly DifficultyRow circleSize;
+        private readonly DifficultyRow drainRate;
+        private readonly DifficultyRow overallDifficulty;
+        private readonly DifficultyRow approachRate;
+        private readonly DifficultyRow stars;
+
+        private readonly Container ratingsContainer;
+        private readonly Bar ratingsBar;
+        private readonly OsuSpriteText negativeRatings;
+        private readonly OsuSpriteText positiveRatings;
+        private readonly BarGraph ratingsGraph;
+
+        private readonly FillFlowContainer retryFailContainer;
+        private readonly BarGraph retryGraph;
+        private readonly BarGraph failGraph;
+
+        private BeatmapInfo beatmap;
+        public BeatmapInfo Beatmap
+        {
+            get
+            {
+                return beatmap;
+            }
+            set
+            {
+                beatmap = value;
+                if (beatmap == null) return;
+
+                description.Text = beatmap.Version;
+                source.Text = beatmap.Metadata.Source;
+                tags.Text = beatmap.Metadata.Tags;
+
+                circleSize.Value = beatmap.Difficulty.CircleSize;
+                drainRate.Value = beatmap.Difficulty.DrainRate;
+                overallDifficulty.Value = beatmap.Difficulty.OverallDifficulty;
+                approachRate.Value = beatmap.Difficulty.ApproachRate;
+                stars.Value = (float)beatmap.StarDifficulty;
+
+                if (beatmap.Metric?.Ratings.Count() == 10)
+                {
+                    var ratings = beatmap.Metric.Ratings.ToList();
+                    ratingsContainer.Show();
+
+                    negativeRatings.Text = ratings.GetRange(0, 5).Sum().ToString();
+                    positiveRatings.Text = ratings.GetRange(5, 5).Sum().ToString();
+                    ratingsBar.Length = (float)ratings.GetRange(0, 5).Sum() / ratings.Sum();
+
+                    ratingsGraph.Values = ratings.Select(rating => (float)rating);
+                }
+                else
+                    ratingsContainer.Hide();
+
+                if (beatmap.Metric?.Retries.Count() == 100 && beatmap.Metric?.Fails.Count() == 100)

```

Comment:
ditto here. `Count > 0` as they will still display correctly even when it's not exactly 100.